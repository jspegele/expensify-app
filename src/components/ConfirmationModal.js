import React from 'react';
import Modal from 'react-modal';

const ConfirmationModal = (props) => (
  <Modal
    appElement={document.getElementById('app')}
    isOpen={props.modalOpen}
    onRequestClose={props.handleCloseConfirmationModal}
    contentLabel="Confirm"
    closeTimeoutMS={200}
    className="modal"
    overlayClassName="overlay"
  >
    <div className="modal__body">
      <p>Are you sure you want to remove this transaction?</p>
    </div>
    <div className="modal__actions">
      <button className="button button--tertiary" onClick={props.handleCloseConfirmationModal}>Cancel</button>
      <button className="button button--destructive" onClick={props.handleRemoveExpense}>Remove</button>
    </div>
  </Modal>
);

export default ConfirmationModal;
