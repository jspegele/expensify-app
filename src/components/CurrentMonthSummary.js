import React from 'react';
import { connect } from 'react-redux';
import moment from 'moment';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';

export const CurrentViewSummary = ({ expensesTotal }) => {
  const budgetWord = expensesTotal === 0 ? 'on target' : ( expensesTotal > 0 ? 'under spending' : 'over spending');
  const formattedExpensesTotal = numeral(expensesTotal / 100).format('$0,0.00');
  const totalClass = expensesTotal > 0 ? 'green-text' : 'red-text';
  return (
    <div className="page-header">
      <div className="content-container">
        <h2 className="page-header__title">
          Your {moment().startOf('month').format('MMMM')} cash flow is <span className={totalClass}>{formattedExpensesTotal}</span>. {expensesTotal > 0 && <span> Nice job!</span>}
        </h2>
      </div>
    </div>
  )
};

const mapStateToProps = (state) => {
  const visibleExpenses = selectExpenses(state.expenses, {
    text: '',
    sortBy: 'amount',
    startDate: moment().startOf('month'),
    endDate: moment().endOf('month')
  });
  
  return {
    expensesTotal: selectExpensesTotal(visibleExpenses)
  };
};

export default connect(mapStateToProps)(CurrentViewSummary);