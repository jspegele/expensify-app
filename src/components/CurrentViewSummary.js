import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';

export const CurrentViewSummary = ({ expenseCount, expensesTotal}) => {
  const expenseWord = expenseCount === 1 ? 'transaction' : 'transactions';
  const formattedExpensesTotal = numeral(expensesTotal / 100).format('$0,0.00');
  const totalClass = expensesTotal > 0 ? 'bold green-text' : 'bold';
  console.log(formattedExpensesTotal);
  return (
    <div className="content-container">
      <div className="view-summary">
        <h2 className="view-summary__title">Viewing <span className="bold">{expenseCount}</span> {expenseWord} totalling <span className={totalClass}>{formattedExpensesTotal}</span>.</h2>
        <div className="view-summary__actions">
          <Link className="button" to="/create">Add Transaction</Link>
        </div>
      </div>
    </div>
  )
};

const mapStateToProps = (state) => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);
  
  return {
    expenseCount: visibleExpenses.length,
    expensesTotal: selectExpensesTotal(visibleExpenses)
  };
};

export default connect(mapStateToProps)(CurrentViewSummary);