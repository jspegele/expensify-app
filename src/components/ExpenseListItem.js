import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import { FaCircle } from 'react-icons/fa';
import numeral from 'numeral';

const ExpenseListItem = ({ dispatch, id, description, amount, createdAt, tags = ''}) => (
  <Link className="list-item" to={`/edit/${id}`}>
    <div className="list-item__description">
      <h3 className="list-item__title">{description}</h3>
      <div><span className="list-item__subtitle">{moment(createdAt).format('MMMM Do, YYYY')}</span></div>
    </div>
    <div className="list-item__details">
      <h3 className="list-item__date">{numeral(amount / 100).format('$0,0.00')}</h3>
      <div className="list-item__tags">
      {
        tags.split(',').map((tag, i) => (
          <span className="list-item__tag" id={tag} key={tag}>
            {i > 0 && <FaCircle size=".5rem" />}
            {tag}
          </span>
        ))
      }
      </div>
    </div>
  </Link>
);

export default ExpenseListItem;