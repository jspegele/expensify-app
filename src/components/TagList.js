import React from 'react';
import { FaTimes } from 'react-icons/fa';

class TagList extends React.Component {
  handleRemoveTag = (e) => {
    const updatedTags = this.props.expense.tags.split(',').filter((tag) => tag !== e.target.id).join(',');
    // this.props.startEditExpense(this.props.expense.id, { tags: updatedTags });
    this.props.handleRemoveTag(updatedTags);
  };
  render() {
    let tags = this.props.expense.tags, tagArray = [], tagClasses = '';
    if(tags && tags.length > 0) {
      tagArray = tags.split(',');
      tagClasses = 'form__tags';
    }
    return (
      <div className={tagClasses}>
        {
          tagArray.map((tag, i) => (
            <button
              className="form__tag"
              id={tag}
              key={i}
              type="button"
              title="Remove tag"
              onClick={this.handleRemoveTag}
            >
              {tag}<FaTimes size="1.4rem" />
            </button>
          ))
        }
      </div>
    );
  };
};

export default TagList;