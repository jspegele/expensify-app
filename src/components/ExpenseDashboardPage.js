import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import CurrentViewSummary from './CurrentViewSummary';
import CurrentMonthSummary from './CurrentMonthSummary';

const ExpenseDashboardPage = () => (
  <div>
    <CurrentMonthSummary />
    <ExpenseListFilters />
    <CurrentViewSummary />
    <ExpenseList />
  </div>
);

export default ExpenseDashboardPage;