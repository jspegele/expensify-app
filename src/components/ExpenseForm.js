import React from 'react';
import { connect } from 'react-redux';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import { FaPlus } from 'react-icons/fa';
import { startEditExpense } from '../actions/expenses';
import TagList from './TagList';

class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      type: (props.expense && props.expense.type) ? props.expense.type : 'expense',
      description: props.expense ? props.expense.description : '',
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      note: props.expense ? props.expense.note : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      newTags: '',
      savedTags: props.expense ? props.expense.tags : '',
      calendarFocused: false,
      modalOpen: false,
      edited: false,
      error: ''
    };
  }
  onTypeChange = (e) => {
    const type = e.target.value;
    this.setState(() => ({ type }));
    if (this.state.description && this.state.amount) { this.setState(() => ({ edited: true })); }
  };
  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
    if (this.state.description && this.state.amount) { this.setState(() => ({ edited: true })); }
  };
  onAmountChange = (e) => {
    const amount = e.target.value;
    
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) { this.setState(() => ({ amount })); }
    if (this.state.description && this.state.amount) { this.setState(() => ({ edited: true })); }
  };
  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
    if (this.state.description && this.state.amount) { this.setState(() => ({ edited: true })); }
  };
  onDateChange = (createdAt) => {
    if(createdAt) {
      this.setState(() => ({ createdAt }));
      if (this.state.description && this.state.amount) { this.setState(() => ({ edited: true })); }
    }
  };
  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };
  onTagsChange = (e) => {
    const newTags = e.target.value;
    if (!newTags || newTags.match(/^[\w\,\s]*$/)) {
      if (newTags.indexOf(',,') > -1) {
        this.setState(() => ({ error: 'Invalid character in tags. Please separate tags by a single comma.' }));
      } else if (newTags.indexOf('  ') === -1) {
        this.setState(() => ({
          newTags,
          error: ''
        }));
        if (this.state.description && this.state.amount) { this.setState(() => ({ edited: true })); }
      }
    } else {
      this.setState(() => ({ error: 'Invalid character in tags. Please use only letters and numbers.' }));
    }
  };
  handleAddTags = () => {
    const combinedTags = this.state.savedTags.concat(',' + this.state.newTags).split(',').map((tag) => tag.trim()).filter((tag) => tag.length > 0);
    const tags = combinedTags.filter((tag, i) => combinedTags.indexOf(tag) === i).join(',').toLowerCase();
    this.props.handleAddTags({ tags });
    this.setState(() => ({
      newTags: '',
      savedTags: tags
    }));
  };
  handleRemoveTag = (savedTags) => {
    this.props.startEditExpense(this.props.expense.id, { tags: savedTags });
    this.setState(() => ({ savedTags }))
  };
  onSubmit = (e) => {
    e.preventDefault();

    const combinedTags = this.state.savedTags.concat(',' + this.state.newTags).split(',').map((tag) => tag.trim()).filter((tag) => tag.length > 0);
    const tags = combinedTags.filter((tag, i) => combinedTags.indexOf(tag) === i).join(',').toLowerCase();
    if(!this.state.description || !this.state.amount) {
      this.setState(() => ({ error: 'Please provide description and amount.' }));
    } else {
      this.setState(() => ({ error: '' }));
      this.props.onSubmit({
        type: this.state.type,
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        note: this.state.note,
        createdAt: this.state.createdAt.valueOf(),
        tags: tags
      });
    }
  };
  render() {
    const saveButtonClass = this.state.edited ? "button" : "button button--disabled";
    const saveTagsClass = this.state.newTags.length > 0 ? "save-tags save-tags--active" : "save-tags";
    return (
      <div>
        <form className="form" onSubmit={this.onSubmit}>
          <label htmlFor="transaction-input">Type</label>
          <div className="input-group input-group--radio" id="transaction-input">
            <label className="label--radio">
              <input type="radio" name="transaction-type" value="expense" onChange={this.onTypeChange} checked={this.state.type === 'expense'} /> Expense
            </label>
            <label className="label--radio">
              <input type="radio" name="transaction-type" value="income" onChange={this.onTypeChange} checked={this.state.type === 'income'} /> Income
            </label>
          </div>
          <label htmlFor="description-input">Description</label>
          <input
            className="text-input"
            id="description-input"
            type="text"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <label htmlFor="amount-input">Amount</label>
          <input
            className="text-input"
            id="amount-input"
            type="text"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <label htmlFor="date-input">Date</label>
          <SingleDatePicker
            id="date-input"
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <label htmlFor="note-input">Note</label>
          <textarea
            className="textarea"
            id="note-input"
            placeholder="Add a note for your expense (optional)"
            value={this.state.note}
            onChange={this.onNoteChange}
          >
          </textarea>
          <label htmlFor="tags-input">Tags</label>
          <div className="input-group input-group--tags">
            <input
              className="text-input"
              id="tags-input"
              type="text"
              placeholder="Add tags separated by commas (ex: travel, hawaii, hotel)"
              value={this.state.newTags}
              onChange={this.onTagsChange}
            />
            { 
              this.props.expense && 
              <FaPlus className={saveTagsClass} size="3.2rem" title="Add tags" onClick={this.handleAddTags} />
            }
          </div>
          { this.props.expense && <TagList expense={this.props.expense} handleRemoveTag={this.handleRemoveTag} />}
          { this.state.error && <p className="form__error">{this.state.error}</p>}
          <div>
            <button className={saveButtonClass} type="submit">Save Transaction</button>
          </div>
        </form>
      </div>
    )
  }
}

const mapDispatchToProps = (dispatch) => ({
  startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense))
});

export default connect(undefined, mapDispatchToProps)(ExpenseForm);
