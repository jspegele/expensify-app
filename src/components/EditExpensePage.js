import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import ConfirmationModal from './ConfirmationModal';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
  state = {
    modalOpen: false
  };
  onSubmit = (expense) => {
    this.props.startEditExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  };
  handleAddTags = (expense) => {
    this.props.startEditExpense(this.props.expense.id, expense);
  };
  onRemove = () => {
    console.log('removing');
    this.setState(() => ({ modalOpen: true }));
  };
  handleRemoveExpense = () => {
    this.handleCloseConfirmationModal();
    this.props.startRemoveExpense({ id: this.props.expense.id });
    this.props.history.push('/');
  };
  handleCloseConfirmationModal = () => {
    this.setState(() => ({ modalOpen: false }));
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h2 className="page-header__title">Edit Transaction</h2>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm
            expense={this.props.expense}
            handleAddTags={this.handleAddTags}
            onRemove={this.onRemove}
            onSubmit={this.onSubmit}
          />
        </div>
        <ConfirmationModal
          modalOpen={this.state.modalOpen}
          handleRemoveExpense={this.handleRemoveExpense}
          handleCloseConfirmationModal={this.handleCloseConfirmationModal}
        />
      </div>
    );
  }
};

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find((expense) => expense.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch, props) => ({
  startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
  startRemoveExpense: (data) => dispatch(startRemoveExpense(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);