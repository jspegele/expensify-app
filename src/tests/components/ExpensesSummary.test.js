import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';
//import getVisibleExpenses from '../selectors/expenses';
//import selectExpensesTotal from '../selectors/expenses-total';

test('should render ExpensesSummary with one expense correctly', () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={1} expensesTotal={235} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render ExpensesSummary with multiple expenses correctly', () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={23} expensesTotal={587624} />);
  expect(wrapper).toMatchSnapshot();
});
