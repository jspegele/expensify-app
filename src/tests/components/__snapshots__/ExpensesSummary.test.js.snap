// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary with multiple expenses correctly 1`] = `
<div>
  <p>
    Viewing 
    23
     
    expenses
     totalling 
    $5,876.24
    .
  </p>
</div>
`;

exports[`should render ExpensesSummary with one expense correctly 1`] = `
<div>
  <p>
    Viewing 
    1
     
    expense
     totalling 
    $2.35
    .
  </p>
</div>
`;
